### ISSUE TITLE
Empty lines get added while iterating over values

### ISSUE THREAD

**Issue by KingDuckZ** [JeffGarland/liaw2014#15]
It is entirely possible I'm doing something wrong, but in my code when I iterate (ie {{#items}} syntax) over values, the output has empty lines where the {{#items}} token used to be.

For example, given this input:

    {
      "hello": [
        {"items": [
          {"name": "red"},
          {"name": "green"},
          {"name": "blue"}
        ]},
        {"items": [
          {"name": "yellow"},
          {"name": "black"},
          {"name": "white"}
        ]}
      ],
      "lol": "hahahah"
    }

for this template:

    {{lol}}
    {{lol}}
    {{lol}} and {{lol}}
    -----------------------
    rigahello{{#hello}}
    {{#items}}
    name: "{{name}}"
    {{/items}}
    {{/hello}}

from the online mustache's demo I get this output:

    hahahah
    hahahah
    hahahah and hahahah
    -----------------------
    rigahello
    name: "red"
    name: "green"
    name: "blue"
    name: "yellow"
    name: "black"
    name: "white"

but from my test code I get:

    hahahah
    hahahah
    hahahah and hahahah
    -----------------------
    rigahello
    
    name: "red"
    
    name: "green"
    
    name: "blue"
    
    
    
    name: "yellow"
    
    name: "black"
    
    name: "white"
    [plus two empty lines that github is stripping]
    
in which some lines that became empty after substitution are kept. The following fix

    diff --git a/include/boost/boostache/frontend/stache/grammar_def.hpp b/include/boost/boostache/frontend/stache/grammar_def.hpp
    index 9024fcd..995c4c1 100644
    --- a/include/boost/boostache/frontend/stache/grammar_def.hpp
    +++ b/include/boost/boostache/frontend/stache/grammar_def.hpp
    @@ -30,6 +30,8 @@
     #include <boost/spirit/include/qi_omit.hpp>
     #include <boost/spirit/include/qi_plus.hpp>
     #include <boost/spirit/include/qi_sequence.hpp>
    +#include <boost/spirit/include/qi_optional.hpp>
    +#include <boost/spirit/include/qi_eol.hpp>
     #include <boost/spirit/include/support_argument.hpp>
     
     namespace boost { namespace boostache { namespace frontend { namespace stache
    @@ -50,6 +52,7 @@ namespace boost { namespace boostache { namespace frontend { namespace stache
           qi::char_type char_;
           qi::lexeme_type lexeme;
           qi::lit_type lit;
    +     qi::eol_type eol;
           qi::matches_type matches;
           qi::no_skip_type no_skip;
           qi::omit_type omit;
    @@ -110,6 +113,7 @@ namespace boost { namespace boostache { namespace frontend { namespace stache
              >> (lit('#') | '^')
              >> identifier
              >> "}}"
    +         >> -eol
              ;
     
           section_end =
    @@ -117,6 +121,7 @@ namespace boost { namespace boostache { namespace frontend { namespace stache
              >> '/'
              >> lit(_r1)
              >> "}}"
    +         >> -eol
              ;
     
           partial =

partially addresses the issues, as it eats up the trailing eol in the input if present. However, in cases where the line is not empty (for example "rigahello{{#hello}}" in my input) the trailing eol shold be preserved, and in this case my fix doesn't work anymore:

    hahahah
    hahahah
    hahahah and hahahah
    -----------------------
    rigahelloname: "red"
    name: "green"
    name: "blue"
    name: "yellow"
    name: "black"
    name: "white"

**Comment by mjcaisse** [comment]
I think the online demo here: http://mustache.github.io/#demo  is wrong.

I think the demo here: http://trymustache.com/ produces the correct answer.

There should be an extra newline between the first and second set of colours:

```
hahahah
hahahah
hahahah and hahahah
-----------------------
rigahello
name: "red"
name: "green"
name: "blue"

name: "yellow"
name: "black"
name: "white"

```

This is based on my reading of the spec here: https://github.com/mustache/spec/blob/master/specs/sections.yml

**Comment by mjcaisse** [comment]
Resolved with commit 5e8d606c426f4b3043a46e9a1ae50842626a753e.

Reworked the entire parser to not have a skipper by default. Added state context while doing the compile to determine how to handle the blank/eol tokens.
